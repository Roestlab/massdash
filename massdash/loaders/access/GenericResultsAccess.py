"""
massdash/loaders/access/GenericResultsAccess
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

from abc import ABC, abstractmethod
import pandas as pd

# Structs
from ...structs.TransitionGroupFeature import TransitionGroupFeature
# Utils
from ...util import LOGGER

class GenericResultsAccess(ABC):
    def __init__(self, filename: str, verbose: bool = False) -> None:
        LOGGER.name = __class__.__name__
        if verbose:
            LOGGER.setLevel("DEBUG")
        else:
            LOGGER.setLevel("INFO")

    @abstractmethod
    def getTransitionGroupFeatures(self, runname: str, pep: str, charge: int) -> TransitionGroupFeature:
        pass

    @abstractmethod
    def getTransitionGroupFeaturesDf(self, runname: str, pep: str, charge: int) -> pd.DataFrame:
        pass

    @abstractmethod
    def getTopTransitionGroupFeature(self, runname: str, pep: str, charge: int) -> TransitionGroupFeature:
        pass


