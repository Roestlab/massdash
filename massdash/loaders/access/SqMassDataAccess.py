"""
massdash/loaders/access/SqMassDataAccess
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

#!/usr/bin/python
# -*- coding: utf-8 -*-
"""
=========================================================================
            MassDash 
=========================================================================

Copyright (c) 2013, ETH Zurich
For a full list of authors, refer to the file AUTHORS.

This software is released under a three-clause BSD license:
 * Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
 * Neither the name of any author or any participating institution
   may be used to endorse or promote products derived from this software
   without specific prior written permission.
--------------------------------------------------------------------------
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
--------------------------------------------------------------------------
$Maintainer: Justin Sing$
$Authors: Hannes Roest, Justin Sing$
--------------------------------------------------------------------------
"""

from typing import List
from collections import OrderedDict
import pyopenms as po
import sqlite3
import pandas as pd
import base64
import struct
import zlib

# Structs
from ...structs.Chromatogram import Chromatogram
# Utils
from ...util import check_sqlite_column_in_table, check_sqlite_table

class SqMassDataAccess:

    def __init__(self, filename):
        self.conn = sqlite3.connect(filename)
        self.c = self.conn.cursor()
        self.filename = filename

    def getPrecursorChromIDs(self, precursor_id):
        """
        Get the chromatogram IDs for a given precursor ID
        """
        data = [row for row in self.c.execute(f"""SELECT ID, NATIVE_ID FROM CHROMATOGRAM WHERE NATIVE_ID LIKE '{precursor_id}_Precursor%'""")]
        return {"chrom_ids":[d[0] for d in data], "native_ids":[d[1] for d in data]}


    def _getChromatogramsHelper(self, ids: List[str], labels: List[str]):
        """
        Helper function to get chromatogram data, assumes that ids are not empty
        """
        stmt ="SELECT CHROMATOGRAM_ID, COMPRESSION, DATA_TYPE, DATA FROM DATA WHERE CHROMATOGRAM_ID IN ("
        for myid in ids:
            stmt += str(myid) + ","
        stmt = stmt[:-1]
        stmt += ")"

        data = [row for row in self.c.execute(stmt)]
        return self._returnDataForChromatogram(data)
    
    def _getChromatogramsHelperFromNativeIds(self, native_ids: List[str]):
        stmt ="SELECT CHROMATOGRAM_ID, COMPRESSION, DATA_TYPE, DATA FROM DATA INNER JOIN CHROMATOGRAM ON CHROMATOGRAM.ID = CHROMATOGRAM_ID WHERE NATIVE_ID IN ("
        for myid in native_ids:
            stmt += str(myid) + ","
        stmt = stmt[:-1]
        stmt += ")"

        data = [row for row in self.c.execute(stmt)]
        return self._returnDataForChromatogram(data)

    def getDataForChromatogramsDf(self, ids: List[str], labels: List[str]) -> pd.DataFrame:
        '''
        Get chromatogram data as a dataframe
        '''
        if len(ids) == 0:
            return pd.DataFrame(columns=['rt', 'intensity', 'annotation'])

        res = self._getChromatogramsHelper(ids, labels)

        c = []
        for l, val in zip(labels, res.values()):
            c.append(Chromatogram(val[0], val[1], l).toPandasDf())

        if len(c) == 0:
            return pd.DataFrame(columns=['rt', 'intensity', 'annotation'])
        else:
            return pd.concat(c)

    def getDataForChromatograms(self, ids: List[str], labels: List[str]) -> List[Chromatogram]:
        """
        Get 
        Get data from multiple chromatograms chromatogram

        - compression is one of 0 = no, 1 = zlib, 2 = np-linear, 3 = np-slof, 4 = np-pic, 5 = np-linear + zlib, 6 = np-slof + zlib, 7 = np-pic + zlib
        - data_type is one of 0 = mz, 1 = int, 2 = rt
        - data contains the raw (blob) data for a single data array
        """

        if len(ids) == 0:
            return [ [ [0], [0] ] ]

        res = self._getChromatogramsHelper(ids, labels)

        ### Convert to chromatograms
        ### match ids with labels
        c = []
        for l, val in zip(labels, res.values()):
            c.append(Chromatogram(val[0], val[1], l))

        return c

    def getDataForChromatogram(self, myid: str, label: str) -> Chromatogram:
        """
        Get data from a single chromatogram

        - compression is one of 0 = no, 1 = zlib, 2 = np-linear, 3 = np-slof, 4 = np-pic, 5 = np-linear + zlib, 6 = np-slof + zlib, 7 = np-pic + zlib
        - data_type is one of 0 = mz, 1 = int, 2 = rt
        - data contains the raw (blob) data for a single data array
        """

        data = [row for row in self.c.execute("SELECT CHROMATOGRAM_ID, COMPRESSION, DATA_TYPE, DATA FROM DATA WHERE CHROMATOGRAM_ID = %s" % myid )]
        res = list(self._returnDataForChromatogram(data).values())[0]
        return Chromatogram(res[0], res[1], label)

    def getDataForChromatogramFromNativeId(self, native_id):
        """
        Get data from a single chromatogram

        - compression is one of 0 = no, 1 = zlib, 2 = np-linear, 3 = np-slof, 4 = np-pic, 5 = np-linear + zlib, 6 = np-slof + zlib, 7 = np-pic + zlib
        - data_type is one of 0 = mz, 1 = int, 2 = rt
        - data contains the raw (blob) data for a single data array
        """

        data = [row for row in self.c.execute("SELECT CHROMATOGRAM_ID, COMPRESSION, DATA_TYPE, DATA FROM DATA INNER JOIN CHROMATOGRAM ON CHROMATOGRAM.ID = CHROMATOGRAM_ID WHERE NATIVE_ID = %s" % native_id )]
        return list(self._returnDataForChromatogram(data).values())[0]

    def getDataForChromatogramsFromNativeIdsDf(self, native_ids: List[str], labels: List[str]) -> pd.DataFrame:
        '''
        Get chromatogram data as a dataframe
        '''
        if len(native_ids) == 0:
            return pd.DataFrame(columns=['rt', 'intensity', 'annotation'])

        res = self._getChromatogramsHelperFromNativeIds(native_ids)

        c = []
        for l, val in zip(labels, res.values()):
            c.append(Chromatogram(val[0], val[1], l).toPandasDf())

        if len(c) == 0:
            return pd.DataFrame(columns=['rt', 'intensity', 'annotation'])
        else:
            return pd.concat(c)

    def getDataForChromatogramsFromNativeIds(self, native_ids: List, labels: List[str]) -> List[Chromatogram]:
        """
        Get data from multiple chromatograms chromatogram

        - compression is one of 0 = no, 1 = zlib, 2 = np-linear, 3 = np-slof, 4 = np-pic, 5 = np-linear + zlib, 6 = np-slof + zlib, 7 = np-pic + zlib
        - data_type is one of 0 = mz, 1 = int, 2 = rt
        - data contains the raw (blob) data for a single data array
        """

        if len(native_ids) == 0:
            return [ [ [0], [0] ] ]
        
        res = self._getChromatogramsHelperFromNativeIds(native_ids)

        ### Convert to chromatograms
        ### match ids with labels
        c = []
        for l, val in zip(labels, res.values()):
            c.append(Chromatogram(val[0], val[1], l))

        return c

    def _returnDataForChromatogram(self, data):
        # prepare result
        chr_ids = set([chr_id for chr_id, compr, data_type, d in data] )
        res = OrderedDict()
        numpress_config = po.NumpressConfig()
        for i in chr_ids:
            res[i] = [None, None]


        for chr_id, compr, data_type, d in data:
            result = []

            if compr == 1:
                tmp = zlib.decompress(d)
                result = struct.unpack("<%sd" % (len(tmp) // 8), tmp)

            if compr == 5:
                # tmp = [ord(q) for q in zlib.decompress(d)]
                tmp = bytearray( zlib.decompress(d) )
                if len(tmp) > 0:
                    numpress_config.setCompression('linear')
                    po.MSNumpressCoder().decodeNP(base64.b64encode(tmp), result, False, numpress_config)
                else:
                    result = [0]
            if compr == 6:
                # tmp = [ord(q) for q in zlib.decompress(d)]
                tmp = bytearray( zlib.decompress(d) )
                if len(tmp) > 0:
                    numpress_config.setCompression('slof')
                    po.MSNumpressCoder().decodeNP(base64.b64encode(tmp), result, False, numpress_config)
                else:
                    result = [0]

            if len(result) == 0:
                result = [ 0 ]
            if data_type == 1:
                res[chr_id][1] = result
            elif data_type == 2:
                res[chr_id][0] = result
            else:
                raise Exception("Only expected RT or Intensity data for chromatogram")

        return res
    
    def __str__(self):
        return f"SqMassDataAccess(filename={self.filename})"
 
    def __repr__(self):
        return f"SqMassDataAccess(filename={self.filename})"
