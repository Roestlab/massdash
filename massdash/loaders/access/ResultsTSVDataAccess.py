"""
massdash/loaders/access/ResultsTSVDataAccess
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

import pandas as pd
import numpy as np
import re
from typing import Literal

# Loaders
from .GenericResultsAccess import GenericResultsAccess
# Structs
from ...structs.TransitionGroupFeature import TransitionGroupFeature
# Utils
from ...util import LOGGER

class ResultsTSVDataAccess(GenericResultsAccess): 
    ''' Class for generic access to TSV file containing the results, currently only supports DIA-NN tsv files'''

    # static variable
    columnMapping = {
        'OpenSwath':{'ProteinName': 'ProteinId', 'Sequence': 'PeptideSequence', 'FullPeptideName': 'ModifiedPeptideSequence', 'm_score': 'Qvalue', 'mz': 'PrecursorMz', 'Charge': 'PrecursorCharge', 'leftWidth': 'RT.Start', 'rightWidth': 'RT.Stop'},
        'DIA-NN':{'Protein.Ids': 'ProteinId', 'Stripped.Sequence': 'PeptideSequence', 'Modified.Sequence': 'ModifiedPeptideSequence', 'Q.Value': 'Qvalue', 'Precursor.Mz': 'PrecursorMz', 'Precursor.Charge': 'PrecursorCharge', 'Precursor.Quantity': 'Intensity', 'Run':'filename'},
        'DreamDIA':{'protein_name': 'ProteinId', 'sequence': 'PeptideSequence', 'full_sequence': 'ModifiedPeptideSequence', 'qvalue': 'Qvalue', 'SCORE_MZ': 'PrecursorMz', 'SCORE_CHARGE': 'PrecursorCharge', 'filename': 'filename', 'quantification': 'Intensity'}

    }

    def __init__(self, filename: str, results_type: Literal["OpenSWATH", "DIA-NN", "DreamDIA"] = "DIA-NN", verbose: bool = False) -> None:
        super().__init__(filename, verbose)
        self.filename = filename
        self.results_type = results_type
        
        self.peptideHash = self._initializePeptideHashTable()
        self.df = self.loadData()   
        self.runs = self.df['filename'].drop_duplicates()  
        self.has_im = 'IM' in self.df.columns
    
    def loadData(self) -> pd.DataFrame:
        '''
        This method loads the data from self.filename into a pandas dataframe
        '''
        df = pd.read_csv(self.filename, sep='\t')

        # rename column according to column mapping 
        df = df.rename(columns=ResultsTSVDataAccess.columnMapping[self.results_type])
        
        # TODO is this required?
        # Assign dummy Decoy column all 0
        df['Decoy'] = 0
        return df

    def _initializePeptideHashTable(self) -> pd.DataFrame:   
        '''
        Load Peptide and Charge for easy access
        '''
        if self.results_type == "OpenSWATH":
            self.hash_table_columns = ['FullPeptideName', 'Charge', 'filename']
            self.rt_multiplier = 1
        elif self.results_type == "DIA-NN":
            self.hash_table_columns = ['Modified.Sequence', 'Precursor.Charge', 'Run']
            self.rt_multiplier = 60
        elif self.results_type == "DreamDIA":
            self.hash_table_columns = ['full_sequence', 'sequence', 'filename']
            self.rt_multiplier = 1
            
        pepHash = pd.read_csv(self.filename, sep='\t', usecols=self.hash_table_columns)

        return pepHash.rename(columns=ResultsTSVDataAccess.columnMapping[self.results_type])
        
    def getTopTransitionGroupFeature(self, runname: str, pep: str, charge: int) -> TransitionGroupFeature:
        '''
        Loads the top TransitionGroupFeature from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            TransitionGroupFeature: TransitionGroupFeature object containing peak boundaries, intensity and confidence
        '''
        tg = self.getTransitionGroupFeatures(runname, pep, charge)

        if len(tg) == 0:
            return None
        elif len(tg) == 1:
            return tg[0]
        else:
            LOGGER.debug(f"Warning: More than one feature found for {pep} {charge} in {self.filename}, returning the first feature, this can lead to unexcepted behaviour")
            return tg[0]

    def getTransitionGroupFeatures(self, runname: str, peptide:str, charge: int):
        '''
        Loads a PeakFeature object from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            TransitionGroupFeature: TransitionGroupFeature object containing peak boundaries, intensity and confidence
        '''
        runname_exact = self.getExactRunName(runname)

        if runname_exact is None:
            LOGGER.debug(f"Error: No matching runs found for {runname}")
            return []
        else:
            targetPeptide = self.peptideHash[(self.peptideHash['filename'] == runname_exact) & (self.peptideHash['ModifiedPeptideSequence'] == peptide) & (self.peptideHash['PrecursorCharge'] == charge)]

            # return the row indices and add 1 to each index to account for the header row
            rows_to_load = [0] + [idx + 1 for idx in targetPeptide.index.tolist()]
                
            # remove any periods from the peptide sequence i.e. for N terminal modifications
            # i.e. Convert .(UniMod:1)SEGDSVGESVHGKPSVVYR to (UniMod:1)SEGDSVGESVHGKPSVVYR
            peptide_tmp = peptide.replace('.', '')
            LOGGER.debug(f"Loading report for {peptide_tmp} {charge} from {self.filename}")

            if len(rows_to_load)-1 !=0:
                feature_data = pd.read_csv(self.filename, sep='\t', skiprows=lambda x: x not in rows_to_load)
                feature_data = feature_data.rename(columns=ResultsTSVDataAccess.columnMapping[self.results_type])
                LOGGER.debug(f"Found {feature_data.shape[0]} rows from {self.filename} for feature data")

                # Save the chromatogram peak feature from the report using cols 'RT', 'RT.Start', 'RT.Stop', 'Precursor.Quantity', 'Q.Value'
                # Multiply RT by 60 to convert from minutes to seconds
                out = []
                for _, row in feature_data.iterrows():
                    out.append(TransitionGroupFeature(consensusApex=row['RT'] * 60,
                                                      leftBoundary=row['RT.Start'] * self.rt_multiplier,
                                                      rightBoundary=row['RT.Stop'] * self.rt_multiplier,
                                                      areaIntensity=row['Intensity'],
                                                      qvalue=row['Qvalue'],
                                                      consensusApexIM=row['IM'] if self.has_im else None,
                                                      sequence=row['ModifiedPeptideSequence'],
                                                      precursor_charge=row['PrecursorCharge']))
                return out 
            else: # len(row_indices)-1==0:
                LOGGER.debug(f"Error: No feature results found for {peptide_tmp} {charge} in {self.filename}")
                return []

    def getTransitionGroupFeaturesDf(self, runname: str, pep_id: str, charge: int) -> pd.DataFrame:
        '''
        Loads a TransitionGroupFeature object from the results file to a pandas dataframe. Since there is only one feature this is the same as getTopTransitionGroupFeatureDf()
        '''
        return self.getTopTransitionGroupFeatureDf(runname, pep_id, charge)

    def getTopTransitionGroupFeatureDf(self, runname: str, pep_id: str, charge: int) -> pd.DataFrame:
        '''
        Get a pandas dataframe with the top TransitionGroupFeatures found in the results file. Since there is only one feature this is the same as getTransitionGroupFeaturesDf
        
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        
        Returns:
            pd.DataFrame: Dataframe with the TransitionGroupFeatures
        '''
        columns = ['filename', 'leftBoundary', 'rightBoundary', 'areaIntensity', 'qvalue', 'consensusApex', 'consensusApexIntensity', 'precursor_charge', 'sequence']
        if self.has_im:
            columns.append('consensusApexIM')
        runname_exact = self.getExactRunName(runname)
        if runname_exact is None:
            return pd.DataFrame(columns=columns)
        else:
            df = self.df[(self.df['filename'] == runname_exact) & (self.df['ModifiedPeptideSequence'] == pep_id) & (self.df['PrecursorCharge'] == charge)]

            df = df.rename(columns={'RT.Start': 'leftBoundary', 
                                    'RT.Stop': 'rightBoundary', 
                                    'Intensity': 'areaIntensity', 
                                    'RT': 'consensusApex', 
                                    'Qvalue' : 'qvalue',
                                    'PrecursorCharge': 'precursor_charge',
                                    'ModifiedPeptideSequence': 'sequence',
                                    'PrecursorMz': 'precursor_mz'})
            
            if self.has_im:
                df = df.rename(columns={'IM': 'consensusApexIM'})

            df['consensusApex'] = df['consensusApex'] * self.rt_multiplier
            df['leftBoundary'] = df['leftBoundary'] * self.rt_multiplier
            df['rightBoundary'] = df['rightBoundary'] * self.rt_multiplier
            df['consensusApexIntensity'] = np.nan
            return df[columns]

    def getExactRunName(self, run_basename_wo_ext: str) -> str:
        '''
        Returns the run name from the filename
        '''
        matching_runs = self.runs[self.runs.str.contains(run_basename_wo_ext)]
        if len(matching_runs) == 0:
            print("Error: No matching runs found for {run_basename_wo_ext}")
            return None
        elif len(matching_runs) == 1:
            return matching_runs.iloc[0]
        else: ## greater than 1
            print("Warning: More than one run found for {run_basename_wo_ext}, this can lead to unpredicted behaviour")
            return matching_runs.iloc[0]
