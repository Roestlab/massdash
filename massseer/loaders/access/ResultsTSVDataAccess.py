import pandas as pd
from typing import List
from massseer.loaders.access.GenericResultsAccess import GenericResultsAccess
from massseer.structs.TransitionGroupFeature import TransitionGroupFeature

class ResultsTSVDataAccess(GenericResultsAccess): 
    ''' Class for generic access to TSV file containing the results, currently only supports DIA-NN tsv files'''
    def init(self, filename: str) -> None:
        self.filename = filename
        self.df = self.loadData()
        self.runs = self.df['filename'].drop_duplicates()
    
    def loadData(self) -> pd.DataFrame:
        '''
        This method loads the data from self.filename into a pandas dataframe
        '''
        df = pd.read_csv(self.filename, sep='\t')

        # rename columns, assuming this is a DIA-NN file
        df = df.rename(columns={'Protein.Ids': 'ProteinId', 'Stripped.Sequence': 'PeptideSequence', 'Modified.Sequence': 'ModifiedPeptideSequence', 'Q.Value': 'Qvalue', 'Precursor.Mz': 'PrecursorMz', 'Precursor.Charge': 'PrecursorCharge'})
        # Assign dummy Decoy column all 0
        df['Decoy'] = 0

        self.df = df


    def getTransitionGroupFeatures(self, runname: str, pep_id: str, charge: int) -> pd.DataFrame:
        '''
        Loads a PeakFeature object from the results file
        Args:
            pep_id (str): Peptide ID
            charge (int): Charge
        Returns:
            TransitionGroupFeature: TransitionGroupFeature object containing peak boundaries, intensity and confidence
        '''
        runname_exact = self.getExactRunName(runname)

        if runname_exact is None:
            return []
        else:
            results = self.df[(self.df['filename'] == runname_exact) & (self.df['ModifiedPeptideSequence'] == pep_id) & (self.df['PrecursorCharge'] == charge)]
            return ResultsTSVDataAccess.convertDfToTransitionGroupFeatures(results)
        
    def getTransitionGroupFeaturesDf(self, runname: str, pep_id: str, charge: int) -> pd.DataFrame:
        '''
        Loads a TransitionGroupFeature object from the results file to a pandas dataframe
        '''
        columns = ['filename', 'leftBoundary', 'rightBoundary', 'areaIntensity', 'qvalue', 'consensusApex', 'consensusApexIntensity']
        runname_exact = self.getExactRunName(runname)
        if runname_exact is None:
            return pd.DataFrame(columns=columns)
        else:
            # TODO renaming or filtering columns needed?
            return self.df[(self.df['filename'] == runname_exact) & (self.df['ModifiedPeptideSequence'] == pep_id) & (self.df['PrecursorCharge'] == charge)]

    @staticmethod
    def convertDfToTransitionGroupFeatures(df: pd.DataFrame) -> List[TransitionGroupFeature]:
        out = []
        for _, row in df.iterrows():
            out.append(TransitionGroupFeature(consensusApex=row['RT'] * 60,  leftBoundary=row['RT.Start'] * 60, rightBoundary=row['RT.Stop'] * 60, areaIntensity=row['Precursor.Quantity'], qvalue=row['Q.Value']))


    def getExactRunName(self, runname: str) -> str:
        '''
        Returns the run name from the filename
        '''
        matching_runs = self.run[self.run['filename'].str.contains(runname)]
        if len(matching_runs) == 0:
            print("Error: No matching runs found for {runname}")
            return None
        elif len(matching_runs) == 1:
            return matching_runs['filename'].iloc[0]
        else: ## greater than 1
            print("Warning: More than one run found for {runname}, this can lead to unpredicted behaviour")
            return matching_runs['filename'].iloc[0]