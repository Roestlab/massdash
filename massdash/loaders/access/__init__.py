"""
massdash/loaders/access
~~~~~~~~~~~~~~~~~~~~~~~

The :mod:`massdash.loaders.access` subpackage contains the low level access to DIA data files. 
"""

from .GenericResultsAccess import GenericResultsAccess
from .MzMLDataAccess import MzMLDataAccess
from .OSWDataAccess import OSWDataAccess
from .ResultsTSVDataAccess import ResultsTSVDataAccess
from .SqMassDataAccess import SqMassDataAccess
from .TransitionPQPDataAccess import TransitionPQPDataAccess
from .TransitionTSVDataAccess import TransitionTSVDataAccess

__all__ = [ "GenericResultsAccess",
            "MzMLDataAccess",
            "OSWDataAccess",
            "ResultsTSVDataAccess",
            "SqMassDataAccess",
            "TransitionPQPDataAccess",
            "TransitionTSVDataAccess"]