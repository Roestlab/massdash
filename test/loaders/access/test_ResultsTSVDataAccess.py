import pytest
from massdash.loaders.access.ResultsTSVDataAccess import ResultsTSVDataAccess
from massdash.testing.PandasSnapshotExtension import PandasSnapshotExtenstion
from syrupy.extensions.amber import AmberDataSerializer


@pytest.fixture
def access(request):
    if request.param == 'diann':
        return ResultsTSVDataAccess("../../test_data/example_dia/diann/report/test_1_diann_report.tsv", 'DIA-NN')
    if request.param == 'dream':
        return ResultsTSVDataAccess("../../test_data/example_dia/dreamdia/test_dreamdia_report.tsv", 'DreamDIA')

@pytest.fixture
def runname():
    return 'test_raw_1'

@pytest.fixture
def peptide():
    return 'DYASIDAAPEER'

@pytest.fixture
def charge():
    return 2

@pytest.fixture
def snapshot_pandas(snapshot):
    return snapshot.use_extension(PandasSnapshotExtenstion)

@pytest.mark.parametrize("access", ['diann', 'dream'], indirect=True)
def test_loadData(access, snapshot_pandas):
    data = access.loadData()
    assert snapshot_pandas == data

@pytest.mark.parametrize("access", ['diann', 'dream'], indirect=True)
def test_initializePeptideHashTable(access, snapshot_pandas):
    hash_table = access._initializePeptideHashTable()
    assert snapshot_pandas == hash_table

#@pytest.mark.parametrize("access,runname,peptide,charge", [('diann', 'test_raw_1', 'DYASIDAAPEER', 2), ('dream', 'test_raw_1', 'DYASIDAAPEER', 2)], indirect=['access'])
@pytest.mark.parametrize("access,runname,peptide,charge", [('diann', 'test_raw_1', 'DYASIDAAPEER', 2),], indirect=['access'])
def test_getTopTransitionGroupFeature(access, runname, peptide, charge, snapshot):
    feature = access.getTopTransitionGroupFeature(runname, peptide, charge)
    assert snapshot == AmberDataSerializer.object_as_named_tuple(feature)

#@pytest.mark.parametrize("access,runname,peptide,charge", [('diann', 'test_raw_1', 'DYASIDAAPEER', 2), ('dream', 'test_raw_1', 'DYASIDAAPEER', 2)], indirect=['access'])
@pytest.mark.parametrize("access,runname,peptide,charge", [('diann', 'test_raw_1', 'DYASIDAAPEER', 2),], indirect=['access'])
def test_getTransitionGroupFeatures(access, runname, peptide, charge, snapshot):
    features = access.getTransitionGroupFeatures(runname, peptide, charge)
    assert snapshot == [ AmberDataSerializer.object_as_named_tuple(i) for i in features ]

#@pytest.mark.parametrize("access,runname,peptide,charge", [('diann', 'test_raw_1', 'DYASIDAAPEER', 2), ('dream', 'test_raw_1', 'DYASIDAAPEER', 2)], indirect=['access'])
@pytest.mark.parametrize("access,runname,peptide,charge", [('diann', 'test_raw_1', 'DYASIDAAPEER', 2),], indirect=['access'])
def test_getTransitionGroupFeaturesDf(access, runname, peptide, charge, snapshot_pandas):
    df = access.getTransitionGroupFeaturesDf(runname, peptide, charge)
    assert snapshot_pandas == df