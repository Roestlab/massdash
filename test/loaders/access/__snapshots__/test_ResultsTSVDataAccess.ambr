# serializer version: 1
# name: test_getIdentifiedPeptides[diann-None]
  dict({
    'test_raw_1': set({
      'AGAANIVPNSTGAAK',
      'DYASIDAAPEER',
      'IEDQLGEVAQYK',
      'KVVITAPGGNDVK',
    }),
  })
# ---
# name: test_getIdentifiedPeptides[diann-test_raw_1]
  set({
    'AGAANIVPNSTGAAK',
    'DYASIDAAPEER',
    'IEDQLGEVAQYK',
    'KVVITAPGGNDVK',
  })
# ---
# name: test_getIdentifiedPrecursors[diann]
  dict({
    'test_raw_1': set({
      'AGAANIVPNSTGAAK_2',
      'AGAANIVPNSTGAAK_3',
      'DYASIDAAPEER_2',
      'DYASIDAAPEER_3',
      'IEDQLGEVAQYK_2',
      'IEDQLGEVAQYK_3',
      'KVVITAPGGNDVK_2',
      'KVVITAPGGNDVK_3',
    }),
  })
# ---
# name: test_getIdentifiedProteins[diann-None]
  dict({
    'test_raw_1': set({
      'P0C0G6;P0C0G7;P0DB18;Q5XDW3;P0DB19;P68777',
      'Q5XD01;P69949;A2RFE3;B5XKM7;P0DA94;P0DA95;P69951;Q1J7I5;Q1JCN8;Q1JHQ6;Q1JML5;Q48UF7',
      'Q5XD49;A2RFQ4;B5XKI1;P0DA82;P0DA83;P69952;Q1J7N4;Q1JCT6;Q1JHV6;Q1JMR3;Q48UK5;Q8P1W4',
    }),
  })
# ---
# name: test_getIdentifiedProteins[diann-test_raw_1]
  set({
    'P0C0G6;P0C0G7;P0DB18;Q5XDW3;P0DB19;P68777',
    'Q5XD01;P69949;A2RFE3;B5XKM7;P0DA94;P0DA95;P69951;Q1J7I5;Q1JCN8;Q1JHQ6;Q1JML5;Q48UF7',
    'Q5XD49;A2RFQ4;B5XKI1;P0DA82;P0DA83;P69952;Q1J7N4;Q1JCT6;Q1JHV6;Q1JMR3;Q48UK5;Q8P1W4',
  })
# ---
# name: test_getNumIdentifiedPeptides[diann-None]
  dict({
    'test_raw_1': 4,
  })
# ---
# name: test_getNumIdentifiedPeptides[diann-test_raw_1]
  4
# ---
# name: test_getNumIdentifiedPrecursors[diann-None]
  dict({
    'test_raw_1': 8,
  })
# ---
# name: test_getNumIdentifiedPrecursors[diann-test_raw_1]
  8
# ---
# name: test_getNumIdentifiedProteins[diann-None]
  dict({
    'test_raw_1': 3,
  })
# ---
# name: test_getNumIdentifiedProteins[diann-test_raw_1]
  3
# ---
# name: test_getTopTransitionGroupFeature[diann-test_raw_1-DYASIDAAPEER-2]
  TransitionGroupFeature(
    areaIntensity=21663.5,
    consensusApex=2438.7000000000003,
    consensusApexIM=0,
    consensusApexIntensity=None,
    getBoundaries=<bound method TransitionGroupFeature.getBoundaries of -------- TransitionGroupFeature --------
  leftBoundary: 2421.63
  rightBoundary: 2455.77
  areaIntensity: 21663.5
  consensusApex: 2438.7000000000003
  consensusApexIntensity: None
  qvalue: 3.28317e-05
  consensusApexIM: 0
  precursor_mz: None
  precursor_charge: 2
  product_annotations: None
  product_mz: None
  sequence: DYASIDAAPEER
  software: DIA-NN>,
    leftBoundary=2421.63,
    precursor_charge=2,
    precursor_mz=None,
    product_annotations=None,
    product_mz=None,
    qvalue=3.28317e-05,
    rightBoundary=2455.77,
    sequence='DYASIDAAPEER',
    software='DIA-NN',
    toPandasDf="TransitionGroupFeature.toPandasDf(transitionGroupFeatureLst: List[ForwardRef('TransitionGroupFeature')]) -> pandas.core.frame.DataFrame",
  )
# ---
# name: test_getTransitionGroupFeatures[diann-test_raw_1-DYASIDAAPEER-2]
  list([
    TransitionGroupFeature(
      areaIntensity=21663.5,
      consensusApex=2438.7000000000003,
      consensusApexIM=0,
      consensusApexIntensity=None,
      getBoundaries=<bound method TransitionGroupFeature.getBoundaries of -------- TransitionGroupFeature --------
  leftBoundary: 2421.63
  rightBoundary: 2455.77
  areaIntensity: 21663.5
  consensusApex: 2438.7000000000003
  consensusApexIntensity: None
  qvalue: 3.28317e-05
  consensusApexIM: 0
  precursor_mz: None
  precursor_charge: 2
  product_annotations: None
  product_mz: None
  sequence: DYASIDAAPEER
  software: DIA-NN>,
      leftBoundary=2421.63,
      precursor_charge=2,
      precursor_mz=None,
      product_annotations=None,
      product_mz=None,
      qvalue=3.28317e-05,
      rightBoundary=2455.77,
      sequence='DYASIDAAPEER',
      software='DIA-NN',
      toPandasDf="TransitionGroupFeature.toPandasDf(transitionGroupFeatureLst: List[ForwardRef('TransitionGroupFeature')]) -> pandas.core.frame.DataFrame",
    ),
  ])
# ---
